From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Fri, 20 Mar 2026 18:26:15 +0100
Subject: applications: Take care of the snap app name if set for portal app
 ID

The portal app ID in snaps can be either `snap.snap-name` or
`snap.snap-name.app-name`

Origin: https://gitlab.gnome.org/GNOME/gnome-control-center/-/merge_requests/3353
---
 panels/applications/cc-applications-panel.c | 42 ++++++++++++++++++++++++-----
 1 file changed, 36 insertions(+), 6 deletions(-)

diff --git a/panels/applications/cc-applications-panel.c b/panels/applications/cc-applications-panel.c
index fbaddea..3f597fa 100644
--- a/panels/applications/cc-applications-panel.c
+++ b/panels/applications/cc-applications-panel.c
@@ -282,12 +282,38 @@ get_portal_app_id (GAppInfo *info)
 
         snap_name = g_desktop_app_info_get_string (G_DESKTOP_APP_INFO (info), "X-SnapInstanceName");
         if (snap_name != NULL)
-            return g_strdup_printf ("%s%s", PORTAL_SNAP_PREFIX, snap_name);
+          {
+            g_autofree gchar *snap_app_name = NULL;
+
+            snap_app_name = g_desktop_app_info_get_string (G_DESKTOP_APP_INFO (info), "X-SnapAppName");
+
+            if (!snap_app_name || g_str_equal (snap_name, snap_app_name))
+                return g_strconcat (PORTAL_SNAP_PREFIX, snap_name, NULL);
+
+            return g_strconcat (PORTAL_SNAP_PREFIX, snap_name, snap_app_name, NULL);
+          }
     }
 
     return NULL;
 }
 
+static inline char *
+get_snap_name_from_portal_id (const char *portal_app_id)
+{
+  g_autofree char *snap_name = NULL;
+  char *sep;
+
+  if (!g_str_has_prefix (portal_app_id, PORTAL_SNAP_PREFIX))
+    return NULL;
+
+  snap_name = g_strdup (portal_app_id + strlen (PORTAL_SNAP_PREFIX));
+
+  if ((sep = strchr (snap_name, '_')) != NULL)
+    *sep = '\0';
+
+  return g_steal_pointer (&snap_name);
+}
+
 static GFile *
 get_flatpak_app_dir (const gchar *app_id, const gchar *subdir)
 {
@@ -687,9 +713,9 @@ remove_snap_permissions (CcApplicationsPanel *self)
 }
 
 static gboolean
-add_snap_permissions (CcApplicationsPanel *self, GAppInfo *info, const gchar *app_id)
+add_snap_permissions (CcApplicationsPanel *self, GAppInfo *info, const gchar *portal_app_id)
 {
-    const gchar *snap_name;
+    g_autofree char *snap_name = NULL;
     g_autoptr(CcSnapdClient) client = NULL;
     g_autoptr(JsonArray) plugs = NULL;
     g_autoptr(JsonArray) slots = NULL;
@@ -697,9 +723,10 @@ add_snap_permissions (CcApplicationsPanel *self, GAppInfo *info, const gchar *ap
     g_autoptr(GError) error = NULL;
     g_autoptr(GError) interfaces_error = NULL;
 
-    if (!g_str_has_prefix (app_id, PORTAL_SNAP_PREFIX))
+    snap_name = get_snap_name_from_portal_id (portal_app_id);
+
+    if (!snap_name)
         return FALSE;
-    snap_name = app_id + strlen (PORTAL_SNAP_PREFIX);
 
     client = cc_snapd_client_new ();
 
@@ -1352,7 +1379,10 @@ update_app_row (CcApplicationsPanel *self, const gchar *app_id)
     g_autofree gchar *formatted_size = NULL;
 
     if (g_str_has_prefix (app_id, PORTAL_SNAP_PREFIX))
-        self->app_size = get_snap_app_size (app_id + strlen (PORTAL_SNAP_PREFIX));
+      {
+        g_autofree char *snap_name = get_snap_name_from_portal_id (app_id);
+        self->app_size = get_snap_app_size (snap_name);
+      }
     else
         self->app_size = get_flatpak_app_size (app_id);
 
